File: lib/ink_flier_web/channels/lobby_channel.ex
defmodule InkFlierWeb.LobbyChannel do
  use InkFlierWeb, :channel
  import TinyMaps

  alias Phoenix.Channel
  alias InkFlier.Lobby
  alias InkFlier.GameServer
  alias InkFlierWeb.Endpoint
  alias InkFlierWeb.GameChannel

  @main_topic "lobby:main"

  def topic, do: @main_topic

  def notify_game_updated(game_id), do: Endpoint.broadcast(@main_topic, "game_updated", GameServer.summary_info(game_id))


  @impl Channel
  def join(@main_topic, _payload, socket), do: {:ok, Lobby.games_info |> Enum.reverse, socket}

  @impl Channel
  def handle_in("create_game", track_id, socket), do: create_game(socket, track_id)

  @impl Channel
  def handle_in("create_and_join_game", track_id, socket), do: create_game(socket, track_id, true)

  @impl Channel
  def handle_in("delete_game", game_id, socket) do
    :ok = Lobby.delete_game(game_id)

    broadcast(socket, "game_deleted", ~M{game_id})
    {:reply, :ok, socket}
  end

  @impl Channel
  def handle_in("join_game", game_id, socket) do
    GameChannel.player_join(game_id, socket.assigns.user)
    {:reply, :ok, socket}
  end

  @impl Channel
  def handle_in("leave_game", game_id, socket) do
    GameChannel.player_leave(game_id, socket.assigns.user)
    {:reply, :ok, socket}
  end


  defp create_game(socket, track_id, join \\ false) do
    {:ok, game_id} = Lobby.start_game(creator: socket.assigns.user, track_id: track_id, join: join)

    broadcast(socket, "game_created", GameServer.summary_info(game_id))
    {:reply, :ok, socket}
  end
end

File: assets/js/lobby_channel.mjs
import {Socket} from "phoenix"

const NEW_GAME = "NEW_GAME"


function create_game(track_id) { channel.push("create_game", track_id) }

function create_and_join_game(track_id) { channel.push("create_and_join_game", track_id) }

function delete_game(gameId) { channel.push("delete_game", gameId) }


function appendGame(gameWrapper) {
  const game_row = makeGameRow(gameWrapper, NEW_GAME)
  gameContainer.prepend(game_row)
}

function drawGamesFromScratch(games) {
  gameContainer.innerHTML = ""
  games.map( (gameWrapper) => {
    const game_row = makeGameRow(gameWrapper, null)
    gameContainer.appendChild(game_row)
  })
}

function makeGameRow(gameWrapper, newGame) {
  const gameId = gameWrapper.name
  const gameCreator = gameWrapper.creator
  const gamePlayers = gameWrapper.players
  const trackId = gameWrapper.track_id

  const element = document.getElementById("game-template").content.cloneNode(true).firstElementChild

  element.dataset.gameId = gameId
  if (newGame) {
    element.classList.add("games__game--new")
  }

  element.querySelector(".gameLink").href = `/lobby/game/${gameId}`
  element.querySelector(".gameId").innerHTML = gameId
  element.querySelector(".track-id").innerHTML = trackId
  element.querySelector(".gameCreator").innerHTML = sanitize(gameCreator)

  element.querySelector(".join-game").addEventListener("click", _resp => { channel.push("join_game", gameId) })
  element.querySelector(".leave-game").addEventListener("click", _resp => { channel.push("leave_game", gameId) })
  element.querySelector(".delete-game").addEventListener("click", _resp => { delete_game(gameId) })

  // NOTE switch this to another template/clone if it gets more complicated than <span>Player1</span>
  element.querySelector(".gamePlayers").innerHTML = ""
  if (gamePlayers.length == 0) {
    const span = document.createElement("span")
    span.innerHTML = "..."
    element.querySelector(".gamePlayers").appendChild(span)
  } else {
    gamePlayers.map( (player) => {
      const span = document.createElement("span")
      span.innerHTML = sanitize(player)
      element.querySelector(".gamePlayers").appendChild(span)
    })
  }

  return element
}


let gameContainer = document.querySelector("#game-list")
let socket = new Socket("/socket", {params: {token: window.userToken}})
socket.connect()

let channel = socket.channel("lobby:main", {})
channel.join()
  .receive("ok", games => {
    drawGamesFromScratch(games)
  })
  .receive("error", resp => { console.log("Unable to join", resp) })

channel.on("game_created", gameWrapper => {
  appendGame(gameWrapper)
})

channel.on("game_updated", gameWrapper => {
  const gameRow = makeGameRow(gameWrapper, null)
  document.querySelector(`[data-game-id="${gameWrapper.name}"]`).replaceWith(gameRow)
})

channel.on("game_deleted", payload => {
  let target = document.querySelector(`[data-game-id="${payload.game_id}"]`)
  target.remove()
})


function sanitize(string) {
  return string.replace(/</g,"&lt;")
}

window.create_game = create_game
window.create_and_join_game = create_and_join_game
window.delete_game = delete_game
export default socket
